ceno_zkvm/src/scheme/mock_prover.rs

```diff
@@ -711,24 +712,25 @@ Hints:
             .into_iter()
             .map(|v| v.into())
             .collect_vec();
-        Self::assert_satisfied(cb, &wits_in, programs, challenge, lkm);
+        Self::assert_satisfied(cb, &wits_in, program, challenge, lkm);
     }
 
     pub fn assert_satisfied(
         cb: &CircuitBuilder<E>,
         wits_in: &[ArcMultilinearExtension<'a, E>],
-        programs: &[ceno_emul::Instruction],
+        program: Arc<Program>,
         challenge: Option<[E; 2]>,
         lkm: Option<LkMultiplicity>,
     ) {
-        Self::assert_with_expected_errors(cb, wits_in, programs, &[], challenge, lkm);
+        Self::assert_with_expected_errors(cb, wits_in, program, &[], challenge, lkm);
     }
 
     pub fn assert_satisfied_full(
-        cs: &ZKVMConstraintSystem<E>,
+        cs: &mut ZKVMConstraintSystem<E>,
```

Comment:
it's not very intuive to me for why `cs` need to be mutable, does it nessesary?

Reply:
I am trying to call `assert_satisfied_raw` inside `assert_satisfied_full`. Currently, `assert_satisfied_raw` takes `CircuitBuilder` struct which wants a `mut ConstraintSystem`.

```rs
Self::assert_satisfied_raw(
    &CircuitBuilder {
        cs: op_cs,
        params: cs.params.clone(),
    },
    ...
);
```

To remove the mutable reference, `assert_satisfied_raw` API need to be changed to take `ConstraintSystem` directly and would cause more changes everywhere.

Reply:
I c the reason, it's because `assert_satisfied_raw` api take `CircuitBuilder`.
But `assert_satisfied_full` take ConstainSystem which cause confuse to be mutable. I realized a small change to make it work

```diff
diff --git a/ceno_zkvm/src/scheme/mock_prover.rs b/ceno_zkvm/src/scheme/mock_prover.rs
index 75340435..f5fe7cda 100644
--- a/ceno_zkvm/src/scheme/mock_prover.rs
+++ b/ceno_zkvm/src/scheme/mock_prover.rs
@@ -726,7 +726,7 @@ Hints:
     }

     pub fn assert_satisfied_full(
-        cs: &mut ZKVMConstraintSystem<E>,
+        cs: &ZKVMConstraintSystem<E>,
         mut fixed_trace: ZKVMFixedTraces<E>,
         witnesses: &ZKVMWitnesses<E>,
         pi: &PublicValues<u32>,
@@ -864,7 +864,7 @@ Hints:
         }

         // assert all opcodes and check lk multiplicity
-        for (name, op_cs) in (cs.circuit_css).iter_mut() {
+        for (name, op_cs) in &cs.circuit_css {
             let is_opcode = op_cs.lk_table_expressions.is_empty()
                 && op_cs.r_table_expressions.is_empty()
                 && op_cs.w_table_expressions.is_empty();
@@ -875,7 +875,7 @@ Hints:
             if let Some(witness) = witnesses.get_table_witness(name) {
                 Self::assert_satisfied_raw(
                     &CircuitBuilder {
-                        cs: op_cs,
+                        cs: &mut op_cs.clone(),
                         params: cs.params.clone(),
                     },
                     witness,
```

Reply:
I've resolved this by removing the requirement of CircuitBuilder as it is not necessary.